### PR TITLE
Add transformation matrix to ColladaExport

### PR DESCRIPTION
**Pull request by gonzodepedro**
Signed-off-by: Gonzalo de Pedro <gonzalo@depedro.com.ar>

This adds transformation matrixes to nodes con collada files when exporting

Connects to ignitionrobotics/ign-gazebo#307

### CODE REVIEW

File: graphics/src/ColladaExporter.cc [hunk 1]
```diff
@@ -152,6 +153,14 @@ ColladaExporter::~ColladaExporter()
 //////////////////////////////////////////////////
 void ColladaExporter::Export(const Mesh *_mesh, const std::string &_filename,
     bool _exportTextures)
+{
+  std::vector<math::Matrix4d> empty;
+  this->ColladaExporter::Export(_mesh, _filename, _exportTextures, empty);
```
Comment (iche033) [inline graphics/src/ColladaExporter.cc:158]: use `this->Export`?

File: graphics/include/ignition/common/ColladaExporter.hh [hunk 2]
```diff
@@ -47,6 +49,10 @@ namespace ignition
       public: virtual void Export(const Mesh *_mesh,
           const std::string &_filename, bool _exportTextures = false);
 
+      public: virtual void Export(const Mesh *_mesh,
+          const std::string &_filename, bool _exportTextures,
+          std::vector<math::Matrix4d> &_submeshToMatrix);
```
Comment (iche033) [inline graphics/include/ignition/common/ColladaExporter.hh:54]: `_submeshToMatrix` is not being modified by the function right? Can you use `const &`?

File: graphics/src/ColladaExporter.cc [hunk 3]
```diff
@@ -152,6 +153,14 @@ ColladaExporter::~ColladaExporter()
...
+}
+
+//////////////////////////////////////////////////
+void ColladaExporter::Export(const Mesh *_mesh, const std::string &_filename,
+    bool _exportTextures, std::vector<math::Matrix4d> &_submeshToMatrix)
```
Comment (iche033) [inline graphics/src/ColladaExporter.cc:163]: does the `_submeshToMatrix` size need to match the `_mesh->SubMeshCount()`? If so, we should add a check and print out an error if it does not

**Review by iche033 — COMMENTED** [review]
I was able to export a [double pendulum model](https://github.com/ignitionrobotics/ign-gazebo/blob/master/examples/worlds/log_record_dbl_pendulum.sdf#L80) so I think simple shapes are working. Can you add a test in `ColladaExporter_TEST`? Something like exporting multiple meshes to a dae file then loading it back with ColladaLoader to make sure they are loaded back as submeshes?

**Comment by mjcarroll** [comment]
@osrf-jenkins run tests

File: graphics/src/ColladaExporter_TEST.cc [hunk 4]
```diff
@@ -193,6 +193,120 @@ TEST_F(ColladaExporter, ExportCordlessDrill)
   common::removeAll(pathOut + "/tmp");
 }
 
+/////////////////////////////////////////////////
+TEST_F(ColladaExporter, ExportMeshWithSubmeshes)
+{
+
+  std::string boxFilenameIn = std::string(PROJECT_SOURCE_PATH) +
+      "/test/data/box.dae";
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter_TEST.cc:201]:
```suggestion
  std::string boxFilenameIn = common::joinPaths(PROJECT_SOURCE_PATH,
      "/test/data/box.dae");
```

File: graphics/src/ColladaExporter_TEST.cc [hunk 5]
```diff
@@ -193,6 +193,120 @@ TEST_F(ColladaExporter, ExportCordlessDrill)
...
+
+  common::ColladaLoader loader;
+  const common::Mesh *boxMesh = loader.Load(
+      boxFilenameIn);
+
+  const common::Mesh *drillMesh = loader.Load(
+      std::string(PROJECT_SOURCE_PATH) +
+      "/test/data/cordless_drill/meshes/cordless_drill.dae");
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter_TEST.cc:209]:
```suggestion
  const common::Mesh *drillMesh = loader.Load(
      common::joinPaths(PROJECT_SOURCE_PATH,
      "/test/data/cordless_drill/meshes/cordless_drill.dae"));
```

File: graphics/src/ColladaExporter_TEST.cc [hunk 6]
```diff
@@ -193,6 +193,120 @@ TEST_F(ColladaExporter, ExportCordlessDrill)
...
+  std::string pathOut = common::cwd();
+  common::createDirectories(pathOut + "/" + "tmp");
+
+  common::ColladaExporter exporter;
+  exporter.Export(&outMesh, pathOut + "/tmp/mesh_with_submeshes", true, subMeshMatrix);
+
+  // Check .dae file
+  tinyxml2::XMLDocument xmlDoc;
+  std::string filename = pathOut +
+     "/tmp/mesh_with_submeshes/meshes/mesh_with_submeshes.dae";
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter_TEST.cc:241]:
```suggestion
  std::string pathOut = common::jointPaths(common::cwd(), "tmp"));
  common::createDirectories(pathOut);

  common::ColladaExporter exporter;
  exporter.Export(&outMesh, 
      common::joinPaths(pathOut, "mesh_with_submeshes", true, subMeshMatrix));

  // Check .dae file
  tinyxml2::XMLDocument xmlDoc;
  std::string filename = common::joinPaths(pathOut,
     "mesh_with_submeshes/meshes/mesh_with_submeshes.dae");
```

File: graphics/src/ColladaExporter_TEST.cc [hunk 7]
```diff
@@ -193,6 +193,120 @@ TEST_F(ColladaExporter, ExportCordlessDrill)
...
+
+    std::string matrixStr = nodeXml->FirstChildElement("matrix")->GetText();
+    EXPECT_EQ(matrixStr, fillData.str());
+
+    nodeXml = nodeXml->NextSiblingElement("node");
+   }
+
+  // Reload mesh and compare
+  const common::Mesh *meshReloaded = loader.Load(pathOut +
+       "/tmp/mesh_with_submeshes/meshes/mesh_with_submeshes.dae");
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter_TEST.cc:294]:
```suggestion
  const common::Mesh *meshReloaded = loader.Load(common::joinPaths(pathOut,
       "mesh_with_submeshes/meshes/mesh_with_submeshes.dae"));
```

**Review by mjcarroll — COMMENTED** [review]
Prefer `common::joinPaths` if you can.

File: graphics/include/ignition/common/ColladaExporter.hh [hunk 8]
```diff
@@ -47,6 +49,10 @@ namespace ignition
       public: virtual void Export(const Mesh *_mesh,
           const std::string &_filename, bool _exportTextures = false);
 
+      public: virtual void Export(const Mesh *_mesh,
+          const std::string &_filename, bool _exportTextures,
+          const std::vector<math::Matrix4d> &_submeshToMatrix);
```
Comment (scpeters) [inline graphics/include/ignition/common/ColladaExporter.hh:54]: I forget the ABI rules. Is it ok to add a virtual function if it's at the end of the class declaration?

Comment (chapulina) [reply graphics/include/ignition/common/ColladaExporter.hh:54]: I believe we haven't been adding new virtual functions, even at the end.

I don't think we have any classes subclassing `ColladaExporter`, so I think this could be non-virtual.

File: graphics/src/ColladaExporter.cc [hunk 9]
```diff
@@ -165,14 +183,6 @@ void ColladaExporter::Export(const Mesh *_mesh, const std::string &_filename,
...
   this->dataPtr->filename = unix_filename.substr(beginFilename);
 
-  if (this->dataPtr->materialCount != 0 &&
-      this->dataPtr->materialCount != this->dataPtr->subMeshCount)
-  {
-    ignwarn << "Material count [" << this->dataPtr->materialCount <<
-        "] different from submesh count [" <<
-        this->dataPtr->subMeshCount << "]\n";
-  }
-
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter.cc:175]: Is this check performed somewhere else now?  I can't tell why it was dropped from here.

File: graphics/src/ColladaExporter.cc [hunk 10]
```diff
@@ -507,13 +520,13 @@ int ColladaExporterPrivate::ExportImages(
       tinyxml2::XMLElement *initFromXml =
         _libraryImagesXml->GetDocument()->NewElement("init_from");
       initFromXml->LinkEndChild(_libraryImagesXml->GetDocument()->NewText(
-        imageString.substr(imageString.find("meshes/")+7).c_str()));
+        ("../materials/textures" + imageString.substr(
+          imageString.rfind("/"))).c_str()));
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter.cc:524]: A small readability enhancement.

```
      const auto imageName = imageString.substr(imageString.rfind("/"));
      const auto imagePath =  common::Join({"..", "materials", "textures", imageName}, "/");
      initFromXml->LinkEndChild(_libraryImagesXml->GetDocument()->NewText(imagePath.c_str());
```

File: graphics/src/ColladaExporter.cc [hunk 11]
```diff
@@ -507,13 +520,13 @@ int ColladaExporterPrivate::ExportImages(
...
       imageXml->LinkEndChild(initFromXml);
 
       if (this->exportTextures)
       {
         createDirectories(this->path + this->filename + "/materials/textures");
-
```
Comment (mjcarroll) [inline graphics/src/ColladaExporter.cc:516]: Nit: unnecessary whitespace change.

File: graphics/src/ColladaExporter_TEST.cc (hunk 4 above)
Comment (mjcarroll) [reply graphics/src/ColladaExporter_TEST.cc:201]: Still needed.

File: graphics/src/ColladaExporter.cc (hunk 9 above)
Comment (gonzodepedro) [reply graphics/src/ColladaExporter.cc:175]: This check is no longer required. There is no longer a one to one relationship between submeshes and materials. Submeshes now can reuse materials used by other submeshes.

**Review by mjcarroll — COMMENTED** [review]
A few lingering unresolved bits

**Comment by gonzodepedro** [comment]
> A few lingering unresolved bits

**Review by mjcarroll — APPROVED** [review]
LGTM, thanks for iterating.

File: graphics/src/ColladaExporter_TEST.cc (hunk 4 above)
Comment (chapulina) [reply graphics/src/ColladaExporter_TEST.cc:201]: It looks like this went through unresolved.

I also noticed this test [failing on CI](https://github.com/ignitionrobotics/ign-common/pull/132/checks?check_run_id=1477457840):

```
  [ RUN      ] ColladaExporter.ExportCordlessDrill
  unknown file: Failure
  C++ exception with description "basic_string::_M_construct null not valid" thrown in the test body.
  [  FAILED  ] ColladaExporter.ExportCordlessDrill (3 ms)
  [ RUN      ] ColladaExporter.ExportMeshWithSubmeshes
  unknown file: Failure
  C++ exception with description "basic_string::_M_construct null not valid" thrown in the test body.
  [  FAILED  ] ColladaExporter.ExportMeshWithSubmeshes (3 ms)
```

Comment (mjcarroll) [reply graphics/src/ColladaExporter_TEST.cc:201]: Not sure why this didn't fail in the action here? 

Fix here: https://github.com/ignitionrobotics/ign-common/pull/133
